api/src/main/java/org/openmrs/api/impl/DiagnosisServiceImpl.java

```diff
@@ -0,0 +1,199 @@
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.api.impl;
+
+import org.openmrs.Diagnosis;
+import org.openmrs.Encounter;
+import org.openmrs.Patient;
+import org.openmrs.api.APIException;
+import org.openmrs.api.DiagnosisService;
+import org.openmrs.api.context.Context;
+import org.openmrs.api.db.DiagnosisDAO;
+import org.openmrs.parameter.EncounterSearchCriteria;
+import org.openmrs.parameter.EncounterSearchCriteriaBuilder;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.*;
+@Transactional
+public class DiagnosisServiceImpl extends BaseOpenmrsService implements DiagnosisService {
+	
+	private DiagnosisDAO diagnosisDAO;
+
+	/**
+	 * Saves a diagnosis
+	 *
+	 * @param diagnosis - the diagnosis to be saved
+	 * @return the diagnosis
+	 */
+	@Override
+	public Diagnosis save(Diagnosis diagnosis) {
+		diagnosisDAO.saveDiagnosis(diagnosis);
+		return diagnosis;
+	}
+
+	/**
+	 * Voids a diagnosis
+	 *
+	 * @param diagnosis  - the diagnosis to be voided
+	 * @param voidReason - the reason for voiding the diagnosis
+	 * @return the diagnosis that was voided
+	 */
+	@Override
+	public Diagnosis voidDiagnosis(Diagnosis diagnosis, String voidReason) {
+		return diagnosisDAO.saveDiagnosis(diagnosis);
+	}
+
+	/**
+	 * Gets a diagnosis based on the uuid
+	 *
+	 * @param uuid - uuid of the diagnosis to be returned
+	 * @return diagnosis matching the given uuid
+	 */
+	@Override
+	public Diagnosis getDiagnosisByUuid(String uuid) {
+		return diagnosisDAO.getDiagnosisByUuid(uuid);
+	}
+
+	/**
+	 * Gets diagnoses since date, sorted in reverse chronological order
+	 *
+	 * @param patient  the patient whose diagnosis we are to get
+	 * @param fromDate the date used to filter diagnosis which happened from this date and later
+	 * @return the list of diagnoses for the given patient and starting from the given date
+	 */
+	@Override
+	@Transactional(readOnly = true)
+	public List<Diagnosis> getDiagnoses(Patient patient, Date fromDate) {
+		EncounterSearchCriteria encounterSearchCriteria = new EncounterSearchCriteriaBuilder().setPatient(patient)
+			.setFromDate(fromDate).createEncounterSearchCriteria();
+		List<Encounter> encounters = Context.getEncounterService().getEncounters(encounterSearchCriteria);
+		List<Diagnosis> diagnoses = new ArrayList<Diagnosis>();
+		for (Encounter encounter : encounters) {
+			diagnoses.addAll(encounter.getDiagnoses());
+		}
+		return diagnoses;
+	}
+
+	/**
+	 * Finds the primary diagnoses for a given encounter
+	 *
+	 * @param encounter the encounter whose diagnoses we are to get
+	 * @return the list of diagnoses in the given encounter
+	 */
+	@Override
+	@Transactional(readOnly = true)
+	public List<Diagnosis> getPrimaryDiagnoses(Encounter encounter) {
+		List<Diagnosis> diagnoses = new ArrayList<Diagnosis>();
+		if (encounter != null && !encounter.getVoided()) {
+			for (Diagnosis diagnosis :
+				encounter.getDiagnoses()) {
+				if (diagnosis.getRank() == 2) {
```

Comment:
What does the magic number 2 mean?

Reply:
Magic number? 
No it isn't. I figured that the diagnosis rank would either be PRIMARY or SECONDARY. I just assigned 2 as the highest level of the diagnosis order.
This is a grey area. 
Any clarifications on the rank and order? @dkayiwa

Reply:
while looking at the code, how do i tell that 2 means SECONDARY?

Reply:
@dkayiwa  I have made mention of that in the comments